src/EditorFeatures/Core/Implementation/BraceMatching/AbstractEmbeddedLanguageBraceMatcher.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.BraceMatching
+{
+    // Note: this type could be concrete, but we cannot export IBraceMatcher's for multiple
+    // languages at once.  So all logic is contained here.  The derived types only exist for
+    // exporting purposes.
+    internal abstract class AbstractEmbeddedLanguageBraceMatcher : IBraceMatcher
+    {
+        public async Task<BraceMatchingResult?> FindBracesAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var languageProvider = document.GetLanguageService<IEmbeddedLanguageProvider>();
+            foreach (var language in languageProvider.GetEmbeddedLanguages())
+            {
+                var braceMatcher = language.BraceMatcher;
+                if (braceMatcher != null)
+                {
+                    var result = await braceMatcher.FindBracesAsync(
```

Comment:
Is there a way we could prune the list of language-specific brace matchers, when we know the language (from `/*lang=json*/` for instance)?

Reply:
...maybe...
This goes back into having the API be more like:   ```languageProvider.TryGetEmbeddedLanguage(SemanticModel, token)``` or ```languageProvider.TryGetEmbeddedLanguage(Document, Position, out language)```.  

However, that approach seems like it will either have a bunch of duplication (because you will pass that data to get the language, and then again when you call into the service APIs), *or* we would end up making it so that things were stateful.  i.e. when you asked for the embedded language, it remembered which state is applied to and used that for hte later ops so you wouldn't have to pass redundant data.

--

I'm willing to try that out... but my gut is telling me that that would be somewhat more complex, without really being much better.

Reply:
Note: something i like about the current system is that it's entirely stateless.  Which means that the provider and the languages are effectively singletons.  I like that that means we don't have to worry about that part of the system producing garbage, and can instead just focus on the individual service implementations being good.

Reply:
Makes sense. Thanks